packages/dds/tree/src/shared-tree/treeCheckout.ts

```diff
@@ -508,67 +509,79 @@ export class TreeCheckout implements ITreeCheckoutFork {
 				}
 			}
 		});
-		_branch.on("commitApplied", (data) => {
-			const commit = _branch.getHead();
-			const { change, revision } = commit;
-			let withinEventContext = true;
-
-			const getRevertible = hasSchemaChange(change)
-				? undefined
-				: (onRevertibleDisposed?: (revertible: Revertible) => void) => {
-						if (!withinEventContext) {
-							throw new UsageError(
-								"Cannot get a revertible outside of the context of a commitApplied event.",
-							);
-						}
-						if (this.revertibleCommitBranches.get(revision) !== undefined) {
-							throw new UsageError(
-								"Cannot generate the same revertible more than once. Note that this can happen when multiple commitApplied event listeners are registered.",
-							);
-						}
-						const revertibleCommits = this.revertibleCommitBranches;
-						const revertible: DisposableRevertible = {
-							get status(): RevertibleStatus {
-								const revertibleCommit = revertibleCommits.get(revision);
-								return revertibleCommit === undefined
-									? RevertibleStatus.Disposed
-									: RevertibleStatus.Valid;
-							},
-							revert: (release: boolean = true) => {
-								if (revertible.status === RevertibleStatus.Disposed) {
-									throw new UsageError(
-										"Unable to revert a revertible that has been disposed.",
-									);
-								}
-
-								const revertMetrics = this.revertRevertible(revision, data.kind);
-								this.logger?.sendTelemetryEvent({
-									eventName: TreeCheckout.revertTelemetryEventName,
-									...revertMetrics,
-								});
-
-								if (release) {
-									revertible.dispose();
-								}
-							},
-							dispose: () => {
-								if (revertible.status === RevertibleStatus.Disposed) {
-									throw new UsageError(
-										"Unable to dispose a revertible that has already been disposed.",
-									);
-								}
-								this.disposeRevertible(revertible, revision);
-								onRevertibleDisposed?.(revertible);
-							},
-						};
-
-						this.revertibleCommitBranches.set(revision, _branch.fork());
-						this.revertibles.add(revertible);
-						return revertible;
-					};
-
-			this.events.emit("commitApplied", data, getRevertible);
-			withinEventContext = false;
+		_branch.on("afterChange", (event) => {
+			// The following logic allows revertibles to be generated for the change.
+			// Currently only appends (including merges) and transaction commits are supported.
+			// In both cases they must not be part of an (outer) transaction.
+			if (!_branch.isTransacting()) {
```

Comment:
Unfortunate diff.

There are a handful of changes here:

* Revertibles are now generated by a loop over "newCommits" rather than always a single revertible generated from the head.
* The process is skipped if the branch is in a transaction (that checked used to happen inside of the branch logic)
* We listen to "append" and "replace" of the branch changed event, rather than listening to the "commitApplied" event (which has been deleted)